snapcraft/internal/elf.py

```diff
@@ -335,8 +337,15 @@ def _extract_attributes(self) -> None:  # noqa: C901
                         self.execstack_set = True
                 elif isinstance(segment, elftools.elf.segments.InterpSegment):
                     self.interp = segment.get_interp_name()
-                elif isinstance(segment, elftools.elf.segments.NoteSegment):
-                    for note in segment.iter_notes():
+
+            # We cannot rely on iterating on the notes via the note segment.
+            # Patchelf may move note sections into other segments, and worse
+            # yet, possibly leave the note segment standing with invalid data.
+            # Here we instead iterate over each section and (more) reliably
+            # gather the note information.
+            for section in elf.iter_sections():
+                if isinstance(section, elftools.elf.sections.NoteSection):
```

Comment:
I was curious about what the best canonical way of determining the build ID is.  I think this is the relevant part of the GDB source code:

https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=bfd/opncls.c;h=99097a9e39989e377af70e9450a970b009439e70;hb=HEAD#l1797

It seems to use the following process:

1. Find the section named `.note.gnu.build-id`
2. Read the contents of the section (this presumably makes sure its not a `SHT_NOBITS` section).
3. Look for a note with name "GNU" and type "NT_GNU_BUILD_ID".  The description field of that note is  the build ID.

So you are right to switch over to checking sections, but I think we can switch this to `section = elf.get_section_by_name('.note.gnu-build-id'), and work from there.

Reply:
Good idea, will do!  Thanks!